Add Changesets workflow (#104)
justeattakeaway/pie: README.md

To add a changeset, run this script locally in the root of the project:

```bash
yarn changeset
```

Follow the prompts to select which package(s) are affected by your change, and whether the change is a major, minor or patch change. This will create a file in the `.changesets` directory of the repo. This change should be committed and included with your PR.

Considerations:

- You can use markdown in your changeset to include code examples, headings, and more. However, **please use plain text for the first line of your changeset**. The formatting of the GitHub release notes does not support headings as the first line of the changeset.
- When selecting packages for the changesets, only select packages which are intended to be published.

### Stable Versions - 'latest' tag

If your change is intended to be released under the `latest` tag on npm, you must follow this workflow:

- Create a branch with your changes. These changes should exclude any `package.json` or manual `CHANGELOG` updates – only include the `.changesets` changes added by Changesets.
- When you create your PR, target the `main` branch.

> - Upon merging to `main`, a new PR titled **Version Packages** is automatically created. This PR includes the `CHANGELOG.md` and `package.json` version bump. Merging this PR will commit this changes to `main` and execute a publish to npm under the `latest` tag.

Reviewer: ```suggestion
- Upon merging to `main`, a new PR titled **Version Packages** is automatically created. This PR includes the `CHANGELOG.md` and `package.json` version bump. Merging this PR will commit these changes to `main` and execute a publish to npm under the `latest` tag.
```

> #### Beta Versions - 'beta' tag
> 
> A Beta release is a release that contains experimental changes. These are ready for early adoption and testing by consumers but may introduce bugs (or be considered work-in-progress).
> 
> #### Feature versions - 'next' tag

Reviewer: Looking at this again, I think these to be the same level as stable versions heading (so h3 too?)

```suggestion
### Beta Versions - 'beta' tag

A Beta release is a release that contains experimental changes. These are ready for early adoption and testing by consumers but may introduce bugs (or be considered work-in-progress).

### Feature versions - 'next' tag
```

Reviewer: Or you could have a h3 which is called Beta and Feature Release above this title, which might then make sense with the `#### Promoting to stable` title below

```

[Yarn docs reference for this](https://yarnpkg.com/getting-started/qa#how-to-share-scripts-between-workspaces)

### Versioning / Publishing Packages

If you are contributing a user-facing or noteworthy change to a pie-monorepo package that should be added to the changelog, you should include a changeset with your PR.

Changesets are only required for Major, Minor, and patch changes that have an effect on consumers. Changes that don't affect consumers do not need to be versioned. These include: linting, testing, CI.

To add a changeset, run this script locally in the root of the project:

```bash
yarn changeset
```

> Follow the prompts to select which package(s) are affected by your change, and whether the change is a major, minor or patch change. This will create a file in the `.changesets` directory of the repo. This change should be committed and included with your PR.

Author: Monorepo root. This directory already exists as it's where the changesets config is stored. When changesets are generated, they get added to this directory 👍🏼

Considerations:

- You can use markdown in your changeset to include code examples, headings, and more. However, **please use plain text for the first line of your changeset**. The formatting of the GitHub release notes does not support headings as the first line of the changeset.
- When selecting packages for the changesets, only select packages which are published.

#### Stable Versions - 'latest' tag

If your change is intended to be released under the `latest` tag on npm, you must follow this workflow:

> - Create a branch with your changes.

Reviewer: ```suggestion
- Create a branch with your changes. These changes should exclude any `package.json` or manual `CHANGELOG` updates – only include the `.changesets` changes added by Changesets.
```

- When you create your PR, target the `main` branch.

> Upon merging to `main`, the CHANGELOG.md, package.json version bump, and npm publish will execute automatically by GitHub Actions.

Reviewer: Just a quick one – I thought once changes were merged to `main`, changesets would open a new PR where the updates would need to be merged from? (so multiple changes would get released together then)

Author: Ah yes, will amend this!

##### Beta Versions - 'beta' tag.

> A Beta release is a release that contains experimental changes that're ready for early adoption by consumers, but may introduce bugs.

Reviewer: ```suggestion
A Beta release is a release that contains experimental changes. These are ready for early adoption and testing by consumers but may introduce bugs (or be considered work-in-progress).
```

##### Feature versions - 'next' tag.

> A Feature release is for larger changes that may require multiple PR's before a version is released. These changes are unstable, and are not intended to be used by consumers. Typically, these releases will be used for testing changes in consuming applications as an alternative to using something like `yalc`.

Reviewer: ```suggestion
A Feature release is for larger changes that may require multiple PRs, across several packages, before it is released. These changes are unstable and are not intended to be used by consumers. Typically, these releases will be used for testing changes in consuming applications as an alternative to using something like `yalc`.
```

If your change is intended to be released under the `next` / `beta` tag on npm, you must follow this workflow:

> - Create a new branch with the `feature-*` / `beta-*` prefix, and push this to the remote.

Reviewer: ```suggestion
- Create a new branch with the `feature-*` / `beta-*` prefix, and push this to the remote. E.g. `git push origin feature-myawesomework`.
```

> - Create another branch to implement your code changes. Ensure this **does not** use a prefix.

Reviewer: ```suggestion
- Create another branch, off this initial feature/beta branch, to implement your code changes. Ensure that this branch **does not** use a prefix.
```

- When you create your PR, target the `feature-*` / `beta-*` branch. 
- GitHub actions will automatically create a new PR that includes the `package.json` version bump, as well as changelog entries.
- Once this generated PR gets merged, GitHub actions will publish your changes under the `next` / `beta` tag.

> **Notes:** Any new PR's that target the `feature-*` / `beta-*` branch will cause GitHub actions to include the changes as part of the release.

Reviewer: ```suggestion
**Notes:** Any new PRs that target the `feature-*` / `beta-*` branch will cause GitHub actions to include the changes as part of that beta/feature release.
```

> Any changes to beta package **must** follow this workflow. PR's that combine changes in `latest` and `beta` packages will result in the beta package being versioned incorrectly.

Reviewer: ```suggestion
Any changes to a beta package **must** follow this workflow. PRs that combine changes in `latest` and `beta` packages will result in the beta package being versioned incorrectly.
```

Reviewer: This line could probably do with a bit more of an explanation. Do you basically mean, "don't merge main into your feature branch work, only rebase the generated PR"?

> ##### Promoting to stable

Reviewer: ```suggestion
#### Promoting to stable
```

When you're happy your `next` / `beta` tagged package is ready to be promoted to a `latest` release, you must use the following workflow.

- Create a PR to merge the `feature-*` / `beta-*` into `main`.

> - Upong merging, a new `latest` release will be available.

Reviewer: ```suggestion
- Upon merging, a new `latest` release will be available.
```

> - Upon merging to `main`, the CHANGELOG.md, package.json version bump, npm publish will execute automatically.

Reviewer: ```suggestion
- Upon merging to `main`, the `CHANGELOG.md`, `package.json` version bump, `npm publish` will execute automatically.
```
